### PR TITLE
feat: P4PU 318 gen access token

### PR DESCRIPTION
**Pull request by oleksiybozhykntt**
#### List of Changes
added access token generation

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: openapi/pagopa-arc-be.openapi.yaml [hunk 1]
```diff
@@ -37,7 +37,42 @@ paths:
             application/json:
               schema:
                 $ref: '#/components/schemas/ErrorDTO'
-
+  /token/oneidentity:
+    get:
+      tags:
+        - arc auth
```
Comment (and-mora) [inline openapi/pagopa-arc-be.openapi.yaml:43]: i'd tag it differently than login endpoint

File: openapi/pagopa-arc-be.openapi.yaml [hunk 2]
```diff
@@ -37,7 +37,42 @@ paths:
...
+            type: string
+      responses:
+        '200':
+          description: "Token Response"
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/TokenResponse'
+        '401':
+          description: "Wrong or missing function key"
```
Comment (and-mora) [inline openapi/pagopa-arc-be.openapi.yaml:67]: In which use case we respond 401?

File: openapi/pagopa-arc-be.openapi.yaml [hunk 3]
```diff
@@ -37,7 +37,42 @@ paths:
...
+        '429':
+          description: "Too many Requests"
+        '500':
+          description: "Service unavailable"
```
Comment (and-mora) [inline openapi/pagopa-arc-be.openapi.yaml:71]: 500 represents an `Internal Server Error`. Check 502 or 503 if you want to model that error

File: src/main/java/it/gov/pagopa/arc/config/OAuth2LoginConfig.java [hunk 4]
```diff
@@ -1,23 +1,40 @@
...
+                authConfig
+                    .baseUri("/login")
+            )
+            .redirectionEndpoint(redirection ->
+                redirection
+                    .baseUri("/token/*")
+            )
+            .successHandler(customAuthenticationSuccessHandler)
         )
+        .sessionManagement(session -> session.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
```
Comment (and-mora) [inline src/main/java/it/gov/pagopa/arc/config/OAuth2LoginConfig.java:35]: does this session management policy work?

File: src/main/java/it/gov/pagopa/arc/dto/IamUserInfoDTO.java
Comment (and-mora) [inline src/main/java/it/gov/pagopa/arc/dto/IamUserInfoDTO.java:1]: this class is not relevant in this PR. Please remove it

File: src/main/java/it/gov/pagopa/arc/service/AccessTokenBuilderService.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    try {
+      rsaPrivateKey = CertUtils.pemKey2PrivateKey(jwtConfiguration.getAccessToken().getPrivateKey());
+      rsaPublicKey = CertUtils.pemPub2PublicKey(jwtConfiguration.getAccessToken().getPublicKey());
+    } catch (InvalidKeySpecException | NoSuchAlgorithmException | IOException e) {
+      throw new IllegalStateException("Cannot load private and/or public key", e);
+    }
+  }
+
+  public TokenResponse build(){
```
Comment (and-mora) [inline src/main/java/it/gov/pagopa/arc/service/AccessTokenBuilderService.java:39]: There is a violation between layers here. We are injecting a controller model inside the service. Try to isolate the layers and move the building of the response elsewhere

**Review by and-mora — CHANGES_REQUESTED** [review]

File: src/test/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  @Mock
+  private Authentication authentication;
+
+  @Mock
+  private TokenResponse tokenResponse;
+
+  private StringWriter responseWriter;
+
+  @Test
+  public void testConstructorInjection() {
```
Comment (and-mora) [inline src/test/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandlerTest.java:49]: what are you testing here?

**Review by and-mora — CHANGES_REQUESTED** [review]

File: src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java
Comment (and-mora) [inline src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java:1]: all immutable class variables must be private and final

File: src/test/java/it/gov/pagopa/arc/service/AccessTokenBuilderServiceTest.java
Comment (and-mora) [inline src/test/java/it/gov/pagopa/arc/service/AccessTokenBuilderServiceTest.java:1]: do not duplicate the keys among test classes

**Review by and-mora — CHANGES_REQUESTED** [review]

File: src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java
Comment (oleksiybozhykntt) [reply src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java:1]: done

File: src/test/java/it/gov/pagopa/arc/service/AccessTokenBuilderServiceTest.java
Comment (oleksiybozhykntt) [reply src/test/java/it/gov/pagopa/arc/service/AccessTokenBuilderServiceTest.java:1]: fixed by using external configuration class

File: src/main/java/it/gov/pagopa/arc/service/AccessTokenBuilderService.java (hunk 5 above)
Comment (oleksiybozhykntt) [reply src/main/java/it/gov/pagopa/arc/service/AccessTokenBuilderService.java]: fixed by returning String

File: src/test/java/it/gov/pagopa/arc/service/AccessTokenBuilderServiceTest.java
[the remark this answers is not in the document]
Comment (oleksiybozhykntt) [reply src/test/java/it/gov/pagopa/arc/service/AccessTokenBuilderServiceTest.java:1]: renamed testing methods

File: src/test/java/it/gov/pagopa/arc/service/AccessTokenBuilderServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+import java.util.regex.Pattern;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class AccessTokenBuilderServiceTest {
+
+  public static final int EXPIRE_IN = 3600;
+
+  private static final String PRIVATE_KEY= """
```
[the remark this answers is not in the document]
Comment (oleksiybozhykntt) [reply src/test/java/it/gov/pagopa/arc/service/AccessTokenBuilderServiceTest.java]: key generated using openssl

**Review by and-mora — APPROVED** [review]

**Review by and-mora — APPROVED** [review]

**Pull request merged**